### PR TITLE
Add Jpeg2000 parser

### PR DESCRIPTION
**Pull request by staniewzki**
## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:
Implemented parser interface for Jpeg2000 format, based on `dali/image/jpeg2k.cc`

### Affected modules and functionalities:
N/A

### Key points relevant for the review:
N/A

### Tests:

- [x] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/imgcodec/parsers/jpeg2000.cc [hunk 1]
```diff
@@ -13,18 +13,57 @@
...
+  // skipping the rest of the block
+  stream.Skip(block_size - sizeof(block_size));
+}
+
+}  // namespace
+
 ImageInfo Jpeg2000Parser::GetInfo(ImageSource *encoded) const {
+  auto stream = encoded->Open();
+
+  advance_one_block(*stream);  // jp2_sig_type
```
Comment (jantonguirao) [inline dali/imgcodec/parsers/jpeg2000.cc:40]: The previous implementation also validated the block type. I think it'd be good to keep that here too. We could argue that `jp2_sig_type` is already checked in `CanParse`. `jp2_format_type` seems not to be validated though.

File: dali/imgcodec/parsers/jpeg2000.cc [hunk 2]
```diff
@@ -13,18 +13,57 @@
...
+  advance_one_block(*stream);  // jp2_format_type
+
+  // jp2 header starts, we skip:
+  // block size
+  // block type (jp_header_type)
+  // image block size
+  // image block type  (jp_im_header_type)
+  stream->Skip<std::array<uint32_t, 4>>();
```
Comment (jantonguirao) [inline dali/imgcodec/parsers/jpeg2000.cc:48]:
```suggestion
  stream->Skip<uint32_t>(4);
```
would be more natural

**Review by jantonguirao — APPROVED** [review]

**Comment by staniewzki** [comment]
!build

File: dali/imgcodec/parsers/jpeg2000.cc [hunk 3]
```diff
@@ -13,18 +13,61 @@
...
+  stream->Skip<uint32_t>();  // block size skipped
+  DALI_ENFORCE(stream->ReadOne<block_type_t>() == jp2_header_type,
+               "Expected block type 'jp2h'");
+  stream->Skip<uint32_t>();  // image block size skipped
+  DALI_ENFORCE(stream->ReadOne<block_type_t>() == jp2_im_header_type,
+               "Expected block type 'ihdr'");
+
+  const int h = ReadValueBE<uint32_t>(*stream);
+  const int w = ReadValueBE<uint32_t>(*stream);
+  const int c = ReadValueBE<uint16_t>(*stream);
```
Comment (mzient) [inline dali/imgcodec/parsers/jpeg2000.cc:56]:
```suggestion
  const int64_t h = ReadValueBE<uint32_t>(*stream);
  const int64_t w = ReadValueBE<uint32_t>(*stream);
  const int64_t c = ReadValueBE<uint16_t>(*stream);
```
uint32_t can overflow. I doubt we would ever see an image with >=2^31 size, but it's theoretically possible. This is dealing with untrusted data, so we should not assume validity nor even benevolence.

**Review by mzient — APPROVED** [review]

**Comment by staniewzki** [comment]
!build

**Pull request merged**